python/ray/workflow/workflow_state.py

```diff
@@ -0,0 +1,290 @@
+        if self.get_input(task_id) is not None:
+            # This case corresponds to the scenario that the task is a
+            # checkpoint or ref.
+            return
+
+        visited_nodes = set()
+        dag_visit_queue = deque([task_id])
+        while dag_visit_queue:
+            tid = dag_visit_queue.popleft()
+            if tid in visited_nodes:
+                continue
+            visited_nodes.add(tid)
+            self.pending_input_set[tid] = set()
+            for in_task_id in self.in_dependency[tid]:
+                self.reference_set[in_task_id].add(tid)
+                # All upstream deps should already complete here,
+                # so we just check their checkpoints.
+                task_input = self.get_input(in_task_id)
+                if task_input is None:
+                    self.pending_input_set[tid].add(in_task_id)
+                    dag_visit_queue.append(in_task_id)
+            if tid in self.latest_continuation:
+                if self.pending_input_set[tid]:
+                    raise ValueError(
+                        "A task that already returns a continuation cannot be pending."
+                    )
+                # construct continuations, as they are not directly connected to
+                # the DAG dependency
+                self.construct_runtime_deps(self.latest_continuation[tid])
+            elif not self.pending_input_set[tid]:
+                self.append_frontier_to_run(tid)
+
+    def init_context(self, context: WorkflowStepContext) -> None:
+        """Initialize the context of all tasks."""
+        for task_id, option in self.task_options.items():
+            self.task_context.setdefault(
+                task_id,
+                dataclasses.replace(
+                    context, task_id=task_id, checkpoint=option.checkpoint
+                ),
+            )
+
+    def add_task_done_callback(self, task_id: StepID, callback: asyncio.Future) -> None:
+        """Add done task callback of a task."""
+        task_id = self.continuation_root.get(task_id, task_id)
+        output = self.get_input(task_id)
+        if output is not None:
+            callback.set_result(output)
+        elif task_id in self.done_tasks:
+            callback.set_exception(
+                ValueError(
+                    f"Task '{task_id}' is done but neither in memory or in storage "
+                    "could we find its output. It could because its in memory "
+                    "output has been garbage collected and the task did not"
+                    "checkpoint its output."
+                )
+            )
+        else:
+            # The task could be a continuation in the future, so we need to take care of
+            # this scenario  in the workflow executor.
+            self.done_task_callbacks[task_id].append(callback)
+
+    def is_untracked_task(self, task_id: StepID) -> bool:
+        """A task is untracked if
```

Comment:
More comment a about the untracked_task and the purpose of this? I failed to understand this part.

Reply:
this is removed